hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -2082,20 +2082,33 @@ func reconcileCAPICluster(cluster *capiv1.Cluster, hcluster *hyperv1.HostedClust
 	cluster.Annotations = map[string]string{
 		hostedClusterAnnotation: client.ObjectKeyFromObject(hcluster).String(),
 	}
-	cluster.Spec = capiv1.ClusterSpec{
-		ControlPlaneEndpoint: capiv1.APIEndpoint{},
-		ControlPlaneRef: &corev1.ObjectReference{
-			APIVersion: "hypershift.openshift.io/v1alpha1",
-			Kind:       "HostedControlPlane",
-			Namespace:  hcp.Namespace,
-			Name:       hcp.Name,
-		},
-		InfrastructureRef: &corev1.ObjectReference{
-			APIVersion: infraCR.GetObjectKind().GroupVersionKind().GroupVersion().String(),
-			Kind:       infraCR.GetObjectKind().GroupVersionKind().Kind,
-			Namespace:  infraCR.GetNamespace(),
-			Name:       infraCR.GetName(),
-		},
+	if infraCR != nil {
+		cluster.Spec = capiv1.ClusterSpec{
+			ControlPlaneEndpoint: capiv1.APIEndpoint{},
+			ControlPlaneRef: &corev1.ObjectReference{
+				APIVersion: "hypershift.openshift.io/v1alpha1",
+				Kind:       "HostedControlPlane",
+				Namespace:  hcp.Namespace,
+				Name:       hcp.Name,
+			},
+			InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: infraCR.GetObjectKind().GroupVersionKind().GroupVersion().String(),
+				Kind:       infraCR.GetObjectKind().GroupVersionKind().Kind,
+				Namespace:  infraCR.GetNamespace(),
+				Name:       infraCR.GetName(),
+			},
+		}
+	} else {
+		// None platform
+		cluster.Spec = capiv1.ClusterSpec{
+			ControlPlaneEndpoint: capiv1.APIEndpoint{},
+			ControlPlaneRef: &corev1.ObjectReference{
+				APIVersion: "hypershift.openshift.io/v1alpha1",
+				Kind:       "HostedControlPlane",
+				Namespace:  hcp.Namespace,
+				Name:       hcp.Name,
+			},
+		}
```

Comment:
@enxebre @csrwng I wonder if this implies another `Platform` interface method to enable platform implementations to build the CAPI Cluster like we do with the CAPI Infrastructure (`ReconcileCAPIInfraCR()`)?

https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/internal/platform/platform.go#L25

Reply:
Another fix for this is to make `ReconcileCAPIInfraCR` [return an object instead of nil](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/internal/platform/none/none.go#L18), so the `InfrastructureRef` can be populated, but I'm not sure if that's a valid approach (since in the None case, we're not creating any Infrastructure, we'd end up returning and presumably not persisting a generic `Cluster` object?)

Reply:
Related prior discussion along these lines: https://github.com/openshift/hypershift/pull/719#discussion_r758638808

If some platforms don't even imply a CAPI Infra or Cluster resource at all, that to me seems like further justification for a contract more like:

```go
type Platform interface {
  DesiredCAPIInfrastructure(...) (client.Object, error)
  DesiredCAPICluster(capiInfrastructure client.Object...) (client.Object, error)
}
```

Where the call side (hostedcluster controller) handles persistence (if the impl returns anything to persist), and where the implementation may return `nil` for either if they don't require those resources.

Reply:
If there's no infraCR it means there's no CAPI support. So please just return early at the top of the func:

```
if infraCR != nil {
 return nil
}
```

Reply:
> If there's no infraCR it means there's no CAPI support. So please just return early at the top of the func:
> 
> ```
> if infraCR != nil {
>  return nil
> }
> ```
> 
> Even better only run this chunk
> 
> ```
> 	// Reconcile the CAPI Cluster resource
> 	capiCluster := controlplaneoperator.CAPICluster(controlPlaneNamespace.Name, hcluster.Spec.InfraID)
> 	_, err = createOrUpdate(ctx, r.Client, capiCluster, func() error {
> 		return reconcileCAPICluster(capiCluster, hcluster, hcp, infraCR)
> 	})
> ```
> 
> `if infra != nil`
> 
> No need to create the cluster CR at all.

This makes me wonder about the `Platform.ReconcileCAPIInfraCR` interface method. Here are its docs:

```go
// ReconcileCAPIInfraCR is called during HostedCluster reconciliation prior to reconciling the CAPI Cluster CR.
// Implementations should use the given input and client to create and update the desired state of the
// platform infrastructure CAPI CR, which will then be referenced by the CAPI Cluster CR.
// TODO (alberto): Pass createOrUpdate construct instead of client.
ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN,
```

Sounds like we're saying if the platform doesn't support CAPI then `ReconcileCAPIInfraCR` should return `nil`. If that's the case, we should document it (and document the follow on implications — does it mean this controller won't call `CAPIProviderDeploymentSpec`?). Seems like there should be a way to make this more clear. For example, what if CAPI was split out into a different interface, like:

```go
type CAPIPlatform interface {
  ReconcileCAPIInfraCR()
  CAPIProviderDeploymentSpec()
}
```

and in the controller, only call those methods at the appropriate time if the platform implements CAPI support. Then `ReconcileCAPIInfraCR()` returning `nil` would be considered an error.

Thoughts?

Reply:
Yeh that all makes total sense to me as a consolidation step. This is intentionally loose atm so we can identify unknowns, platform peculiarities and better boundaries while introducing https://github.com/openshift/hypershift/pull/760 and having empiric feedback.
I'd suggest we iterate to tie the contract farther and go more rigid with something like what you are suggesting after merging this PR and https://github.com/openshift/hypershift/pull/760.